src/App/Datums.cpp

```diff
@@ -0,0 +1,297 @@
+/***************************************************************************
+ *   Copyright (c) 2015 Stefan Tr�ger <stefantroeger@gmx.net>              *
+ *   Copyright (c) 2015 Alexander Golubev (Fat-Zer) <fatzer2@gmail.com>    *
+ *   Copyright (c) 2024 Ondsel (PL Boyer) <development@ondsel.com>         *
+ *                                                                         *
+ *   This file is part of the FreeCAD CAx development system.              *
+ *                                                                         *
+ *   This library is free software; you can redistribute it and/or         *
+ *   modify it under the terms of the GNU Library General Public           *
+ *   License as published by the Free Software Foundation; either          *
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+# include <string>
+#endif
+
+#include <App/Document.h>
+#include <Base/Exception.h>
+#include <Base/Placement.h>
+
+#include "Datums.h"
+
+#ifndef M_PI
+# define M_PI       3.14159265358979323846
+#endif
+
+using namespace App;
+
+PROPERTY_SOURCE(App::DatumElement, App::GeoFeature)
+PROPERTY_SOURCE(App::Plane, App::DatumElement)
+PROPERTY_SOURCE(App::Line, App::DatumElement)
+PROPERTY_SOURCE(App::Point, App::DatumElement)
+PROPERTY_SOURCE(App::LocalCoordinateSystem, App::GeoFeature)
+
+DatumElement::DatumElement(bool hideRole)
+{
+    ADD_PROPERTY_TYPE ( Role, (""), 0, App::Prop_ReadOnly, "Role of the datum in the local coordinate system." ) ;
+
+    // The role is hidden by default. It is visible only when the datum is in a LCS
+    if (hideRole) {
+        Role.setStatus(Property::Hidden, true);
+    }
+}
+
+DatumElement::~DatumElement() = default;
+
+App::LocalCoordinateSystem* DatumElement::getLCS ()
```

Comment:
Formatting issue, the space should not be there

Reply:
This is old code that has quite a lot of formating exotic things like this. I'll try to fix them.